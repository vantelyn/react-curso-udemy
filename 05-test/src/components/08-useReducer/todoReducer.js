// action = {
//     type: 'add',
//     payload: {
//         id: new Date().getTime(),
//         desc: 'DescripciÃ³n del todo...',
//         done: false
//     }
// };

// action = {
//     type: 'delete',
//     payload: 123456789      //ide del ToDo a eliminar
// };

// action = {
//     type: 'toggle',
//     payload: 123456789      //ide del ToDo a cambiar
// };

export const todoReducer = ( state = [], action ) => {

    switch ( action.type ) {

        case 'add':
            return [ ...state, action.payload ];

        case 'delete':
            return state.filter( todo => todo.id !== action.payload ); // 123123123

        case 'toggle': 
            return state.map( todo => 
                ( todo.id === action.payload )
                    ? { ...todo, done: !todo.done }
                    : todo
            );

        case 'toggle-old':
            return state.map( todo => {

                if ( todo.id === action.payload ) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }

            })

        default:
            return state;
    }


}