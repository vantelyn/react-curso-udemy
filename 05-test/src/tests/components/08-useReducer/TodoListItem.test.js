import { shallow } from "enzyme";
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { demoTodos } from "../../fixtures/demoTodos";

describe('Pruebas en el componente <TodoListItem />', () => { 

    const handleDelete=jest.fn();
    const handleToggle=jest.fn();

    const wrapper = shallow(<TodoListItem 
        todo={demoTodos[0]}
        index={1}
        handleDelete = { handleDelete }
        handleToggle = { handleToggle } 
    />);

    test('debe mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();

    });
    
    test('debe llamar la función borrar', () => { 

        wrapper.find('button').simulate('click');
        expect(handleDelete).toHaveBeenCalledWith(demoTodos[0].id);


    });

    test('debe llamar la función toggle', () => { 

        wrapper.find('p').simulate('click');
        expect(handleToggle).toHaveBeenCalledWith(demoTodos[0].id);

    });

    test('debe mostrar el texto correctamente', () => { 

        const texto = wrapper.find('p').text();
        expect(texto).toBe(`2. ${demoTodos[0].desc}`);

    });

    test('debe mostrar el texto correctamente', () => { 

        const todo = demoTodos[0];
        todo.done = true;
        const wrapper = shallow(<TodoListItem 
            todo={todo}
            index={1}
            handleDelete = { handleDelete }
            handleToggle = { handleToggle } 
        />);

        expect( wrapper.find('p').hasClass('complete') ).toBe(true);
    });
    
});