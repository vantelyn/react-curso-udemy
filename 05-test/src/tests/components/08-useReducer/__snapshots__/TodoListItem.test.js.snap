// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoListItem /> debe mostrarse correctamente 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    2
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
